File: yggy/observable/observable_manager.py
from typing import Any, cast

from ..comm import (
    COMM_ADD_CLIENT_MSG,
    Comm,
    Message,
    STOP_PROPAGATION,
    StopPropagationType,
)
from .messages import (
    OBSERVABLE_CHANGE_MSG,
    OBSERVABLE_CLIENT_CHANGE_MSG,
    OBSERVABLE_READY_MSG,
    OBSERVABLE_REGISTER_MSG,
)
from .observable import Observable, ObservableChangeMessage
from .observable_value import ObservableValue

__all__ = [
    "ObservableManager",
]


class ReadyMessage(Message):
    ...


class ObservableManager:
    __registry: dict[str, Observable[Any]]
    __comm: Comm

    __notifying: set[str]

    def __init__(self, __comm: Comm) -> None:
        self.__registry = {}
        self.__comm = __comm
        self.__notifying = set()

        self.__comm.recv(OBSERVABLE_CLIENT_CHANGE_MSG, self.__recv_change)
        self.__comm.recv(COMM_ADD_CLIENT_MSG, self.__recv_add_client)

    @property
    def comm(self) -> Comm:
        return self.__comm

    def register(self, __observable: Observable[Any]) -> None:
        self.__registry[__observable.id] = __observable

    def unregister(self, __observable: Observable[Any]) -> None:
        del self.__registry[__observable.id]

    def notify_change(self, __change: ObservableChangeMessage[Any]) -> None:
        data_id = __change["data_id"]

        if data_id in self.__notifying:
            return

        try:
            self.__notifying.add(data_id)
            self.__comm.send(OBSERVABLE_CHANGE_MSG, __change)
        finally:
            self.__notifying.discard(data_id)

    def __recv_change(
        self, __change: ObservableChangeMessage[Any]
    ) -> StopPropagationType:
        data_id = __change["data_id"]

        if data_id in self.__registry:
            obs = self.__registry[data_id]

            new_value = __change["new_value"]
            if isinstance(obs, ObservableValue):
                obs = cast(ObservableValue[Any], obs)
                obs.set(new_value)

        return STOP_PROPAGATION

    def __recv_add_client(self, __client_id: str) -> None:
        for observable in self.__registry.values():
            self.__comm.send(
                OBSERVABLE_REGISTER_MSG, observable.__json__(), client_ids=[__client_id]
            )
        self.__comm.send(OBSERVABLE_READY_MSG, ReadyMessage(), client_ids=[__client_id])

File: yggy/observable/observable_manager.ts
import { Comm } from "../comm/comm.js";
import { __uuid4 } from "../utils.js";
import {
    ChangeMessage,
    OBSERVABLE_CHANGE_MSG,
    OBSERVABLE_CLIENT_CHANGE_MSG,
    OBSERVABLE_REGISTER_MSG,
    RegisterMessage,
    RegisterObjectMessage,
    RegisterValueMessage,
} from "./messages.js";
import { Observable } from "./observable.js";
import { ObservableObject } from "./observable_object.js";
import { ObservableValue } from "./observable_value.js";

export class ObservableManager {
    private __registry: { [key: string]: Observable };
    private __updating: Set<string>;
    private __comm: Comm;

    constructor(__comm: Comm) {
        this.__registry = {};
        this.__updating = new Set();
        this.__comm = __comm;

        this.__comm.recv(OBSERVABLE_CHANGE_MSG, this.__recv_change.bind(this));
        this.__comm.recv(
            OBSERVABLE_REGISTER_MSG,
            this.__recv_register.bind(this)
        );
    }

    public get comm(): Comm {
        return this.__comm;
    }

    public get(__id: string | number): Observable | undefined {
        if (typeof __id === "number") {
            return Object.values(this.__registry)[__id];
        }
        return this.__registry[__id];
    }

    public notify_change(__change: ChangeMessage<any>): void {
        if (this.__updating.has(__change.data_id)) {
            return;
        }

        this.__comm.send(OBSERVABLE_CLIENT_CHANGE_MSG, __change);
    }

    private __recv_change(__change: ChangeMessage<any>): void {
        const { data_id } = __change;

        if (this.__updating.has(data_id)) {
            return;
        }
        this.__updating.add(data_id);

        this.__registry[data_id]._recv_change(__change);

        this.__updating.delete(data_id);
    }

    private __recv_register(__observable: RegisterMessage): void {
        if (__observable.observable_type == "value") {
            const { data_id, value } = <RegisterValueMessage<any>>__observable;
            this.__registry[data_id] = new ObservableValue(this, data_id, value);
        }

        if (__observable.observable_type == "object") {
            const { data_id, attrs } = <RegisterObjectMessage>__observable;
            this.__registry[data_id] = new ObservableObject(this, data_id, attrs);
        }
    }

}
